### PR TITLE
full_posix_sockets

### PR DESCRIPTION
**Pull request by juj**
This adds full POSIX sockets API support via a `-s PROXY_POSIX_SOCKETS=1` option. This is implemented by proxying out the sockets API calls to a separate proxy bridge process running on localhost. (that one needs to separately spawn)

This can be useful for special occassions, e.g. when building unit test harnesses or similar, where one just wants to compile existing POSIX sockets code and "have it work" locally. A web browser page can then be simulated to act as a TCP or UDP server, for example.

The feature builds on top of #7670, and can only land after #7670 lands.

Will need splitting out libc to two .bc library parts, one for proxy sockets variant, and another one with the musl stubs.

### CODE REVIEW

**Comment by juj** [comment]
This PR would now be good to land, but this is producing some errors on the upstream bots that I do not understand, and am unable to reproduce. @kripken Does any of that make sense to you?

**Comment by kripken** [comment]
I can reproduce those errors locally, using the LLVM wasm backend. E.g. `other.test_export_all_` fails on 
```
error: undefined symbol: accept
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
error: undefined symbol: bind
error: undefined symbol: connect
error: undefined symbol: getsockname
error: undefined symbol: recv
error: undefined symbol: recvfrom
error: undefined symbol: send
error: undefined symbol: sendto
error: undefined symbol: setsockopt
error: undefined symbol: socket
Error: Aborting compilation due to previous errors
```

**Comment by juj** [comment]
Updated this PR and it now passes. Ready for a second review?

**Comment by kripken** [comment]
I don't understand the motivation here yet. What does "proxy" mean - what is proxied where? Is this related to threading proxying?

**Comment by juj** [comment]
Proxying here means that instead of implementing the POSIX sockets calls in the browser, each sockets API call is serialized out via WebSockets pipe over to a native "websockify" style intermediate process, which natively performs those exact sockets calls, and returns the results back - making the Emscripten page look like it has full sockets API access available to it. In order to pull it off, it does require threading proxying, because this only works in multithreaded PROXY_TO_PTHREAD builds, since the command proxying is synchronous.

**Comment by juj** [comment]
Updated this PR, and it now looks green to merge. Good for a review pass?

File: site/source/docs/porting/networking.rst [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+Full POSIX Sockets over WebSocket Proxy Server
+==============================================
+
+Emscripten provides a native POSIX Sockets proxy server program, located in directory tools/websocket_to_posix_proxy/, that allows full POSIX Sockets API access from a web browser. This support works by proxying all POSIX Sockets API calls from the browser to the Emscripten POSIX Sockets proxy server (via transparent use of WebSockets API), and the proxy server then performs the native TCP/UDP calls on behalf of the page. This allows a web browser page to run full TCP & UDP connections, act as a server to accept incoming connections, and perform host name lookups and reverse lookups. Because all API calls are individually proxied, this support can be slow. This support is mostly useful for developing testing infrastructure and debugging.
+
+For an example of how the POSIX Sockets proxy server works in an Emscripten client program, see the file tests/websocket/tcp_echo_client.cpp.
+
```
Comment (kripken) [inline site/source/docs/porting/networking.rst]: The docs don't seem to mention the new proxy option from this PR?

File: system/lib/libc.symbols [hunk 2]
```diff
@@ -1,310 +1,541 @@
--------- C h_errno
--------- C __hwcap
```
Comment (kripken) [inline system/lib/libc.symbols]: why is this diff so large? I think the symbols should be sorted - maybe you didn't sort them before applying?

File: tools/system_libs.py [hunk 3]
```diff
@@ -19,6 +19,11 @@
 stdout = None
 stderr = None
 
+LIBC_SOCKETS = ['socket.c', 'socketpair.c', 'shutdown.c', 'bind.c', 'connect.c',
+                'listen.c', 'accept.c', 'getsockname.c', 'getpeername.c', 'send.c',
+                'recv.c', 'sendto.c', 'recvfrom.c', 'sendmsg.c', 'recvmsg.c',
+                'getsockopt.c', 'setsockopt.c', 'freeaddrinfo.c']
```
Comment (kripken) [inline tools/system_libs.py]: these are included in those builds - so what is proxied here, the syscalls called by those?

**Comment by juj** [comment]
> Thanks, so this lets you use normal C sockets code, and it synchronously proxies commands to a websockets server that does the actual sockets operations? And you say it's useful for things like test harnesses etc. - this seems like quite a lot of work for just that? I feel like I'm missing the use case.

We use this to target an existing sockets utilizing debug suite over to HTML5, retaining the same codebase for portability. This lets one reuse existing POSIX sockets code as is, and one only needs to add one API call at the start of the program to the bridge server, and then you get full sockets. This way we are able to locally connect a Unity page to all sorts of existing harness tools that are running on local developer's machine (e.g. debugger, profiler, inspector)

File: system/lib/libc.symbols (hunk 2 above)
Comment (juj) [reply system/lib/libc.symbols]: This is an as-is result from calling llvm-nm on new generated libc.bc. Is there some special process that one must do instead of dumping them? Perhaps there should be an automatic script that updates the symbol files from already built cache?

File: site/source/docs/porting/networking.rst (hunk 1 above)
Comment (juj) [reply site/source/docs/porting/networking.rst]: Updated docs with linker settings for each case.

File: tools/system_libs.py (hunk 3 above)
Comment (juj) [reply tools/system_libs.py]: These different API calls are proxied through a WebSocket server. I.e. instead of socket() calling musl which calls socketcall(), the musl functions are replaced with C implemented functions that proxy each call through the WebSockets bridge.

File: system/lib/libc.symbols (hunk 2 above)
Comment (kripken) [reply system/lib/libc.symbols]: We should probably add a script, but basically just pipe it through ` | sort`

File: embuilder.py [hunk 4]
```diff
@@ -54,11 +54,13 @@
     'gl-webgl2',
     'html5',
     'libc',
+    'libc-sockets',
```
Comment (kripken) [inline embuilder.py]: the name may be misleading, since it's not the normal sockets code. how about `sockets-proxy` or `sockets-testharness` or such?

Comment (juj) [reply embuilder.py]: The library `libc-sockets` does contain the normal sockets code from musl. These were split off to a separate block so that they can be omitted from the build when the proxy code is used. The proxy sockets code is contained in library `libc-sockets-proxy`. (Renamed to this from `libposix-sockets-proxy` for consistency) If proxy code is enabled, `libc-sockets-proxy` is used, otherwise `libc-sockets` is used.

File: system/lib/libc.symbols (hunk 2 above)
Comment (juj) [reply system/lib/libc.symbols]: Ah gotcha, updated the symbols files (using embuilder to build them, then `llvm-nm ~/.emscripten_cache/asmjs/libc.bc | sort > system/lib/libc.symbols`

**Comment by jakogut** [comment]
This is exactly what I need! I'm building and testing now.

**Comment by jakogut** [comment]
> Thanks, so this lets you use normal C sockets code, and it synchronously proxies commands to a websockets server that does the actual sockets operations? And you say it's useful for things like test harnesses etc. - this seems like quite a lot of work for just that? I feel like I'm missing the use case.

@kripken For testing, this might be a lot of work when there are easier alternatives, but there are other interesting use cases. The proxy would allow unmodified apps to be compiled and use standard networking APIs, without requiring servers that understand WebSockets.

This obviously doesn't help when the proxy isn't running, but embedded applications could include the proxy to allow native networking through the browser. It would be interesting, for instance, to build an embedded Linux system with Chromium and a socket proxy, and serve unmodified native apps through Chromium with full network access.

File: tools/websocket_to_posix_proxy/src/posix_sockets.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <stdio.h>
+
+#if defined(__APPLE__) || defined(____linux__)
```
Comment (jakogut) [inline tools/websocket_to_posix_proxy/src/posix_sockets.h]: Should this be `__linux__`?

File: tools/websocket_to_posix_proxy/src/main.cpp [hunk 6]
```diff
@@ -0,0 +1,382 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <memory.h>
+#include <sys/types.h>
+#include "posix_sockets.h"
+#include "threads.h"
+#include <assert.h>
+#include <vector>
+
```
Comment (jakogut) [inline tools/websocket_to_posix_proxy/src/main.cpp]: It seems that `signal.h` is missing here for at least `__linux__`. I assume `__APPLE__` uses the same header, but I don't know.

File: tools/websocket_to_posix_proxy/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,13 @@
+cmake_minimum_required(VERSION 2.8)
+
+project(websocket_to_posix_proxy)
+
+file(GLOB sourceFiles src/*.cpp src/*.h)
+
+add_executable(websocket_to_posix_proxy ${sourceFiles})
+
```
Comment (jakogut) [inline tools/websocket_to_posix_proxy/CMakeLists.txt]: Running the tests, I encountered:

> [100%] Linking CXX executable websocket_to_posix_proxy
/usr/bin/ld: CMakeFiles/websocket_to_posix_proxy.dir/src/main.cpp.o: in function `main':
main.cpp:(.text+0x1144): undefined reference to `pthread_create'

On Linux, this fixed the issue.

```suggestion
find_package(Threads)
target_link_libraries(websocket_to_posix_proxy ${CMAKE_THREAD_LIBS_INIT})
```

File: src/library.js [hunk 8]
```diff
@@ -3608,6 +3608,7 @@ LibraryManager.library = {
   // arpa/inet.h
   // ==========================================================================
 
+#if PROXY_POSIX_SOCKETS == 0
```
Comment (jakogut) [inline src/library.js]: Running `test/runner.py sockets.test_posix_proxy_sockets` yields:
> error: undefined symbol: inet_addr

File: site/source/docs/porting/networking.rst [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+Full POSIX Sockets over WebSocket Proxy Server
+==============================================
+
+Emscripten provides a native POSIX Sockets proxy server program, located in directory tools/websocket_to_posix_proxy/, that allows full POSIX Sockets API access from a web browser. This support works by proxying all POSIX Sockets API calls from the browser to the Emscripten POSIX Sockets proxy server (via transparent use of WebSockets API), and the proxy server then performs the native TCP/UDP calls on behalf of the page. This allows a web browser page to run full TCP & UDP connections, act as a server to accept incoming connections, and perform host name lookups and reverse lookups. Because all API calls are individually proxied, this support can be slow. This support is mostly useful for developing testing infrastructure and debugging.
+
+To use POSIX sockets proxying, link the application with flags "-lwebsocket.js -s PROXY_POSIX_SOCKETS=1 -s USE_PTHREADS=1 -s PROXY_TO_PTHREAD=1". That is, POSIX sockets proxying builds on top of the Emscripten WebSockets library, and requires multithreading and proxying the application main() to a pthread.
```
Comment (jakogut) [inline site/source/docs/porting/networking.rst]: It's possible `FORCE_FILESYSTEM=1` may also be required to prevent required code from being optimized out.

**Comment by jakogut** [comment]
It seems `PROXY_POSIX_SOCKETS` is incompatible with the SDL2 port at the moment, as `PROXY_TO_PTHREAD` causes http://localhost:6931/test.worker.js:182: Uncaught ReferenceError: screen is not defined`.

**Comment by kripken** [comment]
@jakogut - thanks for the info! One of my concerns in this PR is that I wasn't sure this would have additional users. And thanks for reviewing, that will help move this along.

**Comment by jakogut** [comment]
@kripken Happy to help in any way I can. I also submitted a PR to @juj with some proposed fixes for the issues I encountered.

**Comment by jakogut** [comment]
@juj There appears to be a problem with a program I'm attempting to compile that uses`getpeername`. Upon being called, the proxy server reports: `Received too small sockets call message! size: 0 bytes, expected at least 8 bytes`, and upon investigating, the size passed to `emscripten_websocket_send_binary` is negative. I'm looking into it.

**Comment by jakogut** [comment]
@juj A library I'm trying to use appears to `poll` a socket before reading from it, and enters an infinite loop waiting for data. If I understand correctly, a `poll` would not currently be proxied, would it?

File: system/lib/websocket/websocket_to_posix_socket.cpp [hunk 10]
```diff
@@ -0,0 +1,965 @@
...
+    int address_len;
+    uint8_t address[];
+  };
+
+  PosixSocketCallResult *b = allocate_call_result(sizeof(Result));
+  d.header.callId = b->callId;
+  d.header.function = POSIX_SOCKET_MSG_GETPEERNAME;
+  d.socket = socket;
+  d.address_len = *address_len;
+  emscripten_websocket_send_binary(bridgeSocket, &d, sizeof(d) + *address_len - MAX_SOCKADDR_SIZE);
```
Comment (jakogut) [inline system/lib/websocket/websocket_to_posix_socket.cpp]:
```suggestion
  emscripten_websocket_send_binary(bridgeSocket, &d, sizeof(d);
```

I'm not sure why the size passed to `emscripten_websocket_send_binary` has `address_len` added and `MAX_SOCKADDR_SIZE` subtracted, but removing that fixes the issue I described in this comment. https://github.com/emscripten-core/emscripten/pull/7672#issuecomment-470742919

**Comment by juj** [comment]
Sorry for the long delay in getting back to this PR @jakogut, I have been away from work for a while. Really cool to see this feature have interest for other users as well!

I am rebasing this PR on top of latest incoming to resolve errors - I'll pick up some of the fixes you posted while at it. Dropped comments to https://github.com/juj/emscripten/pull/2/ for other parts.

**Comment by juj** [comment]
> @juj Would it be feasible to make full posix sockets work without `PROXY_TO_PTHREAD`? Maybe starting a separate thread for proxying socket calls?

The main limitation is that main web browser thread cannot synchronously block, or web browser JavaScript watchdog timer will abort web page execution. Also, since JS events are processed on the main browser thread, it means that nothing "moves forward" while main browser thread is executing JS, so main thread cannot block even for things that might have an expected quick turnaround time, since that turnaround cannot happen while main thread is busy. Only when main thread JS yields back to the web browser to continue processing, will events tick foward.

Proxied posix sockets should work in pthreads when `PROXY_TO_PTHREAD` is not enabled, as long as the main thread does not perform any sockets calls. Performance of proxying could be improved by offloading the proxying processing to a dedicated Worker thread, if main browser thread is regularly busy executing some other user JS/Wasm code.

This means that even if there was a dedicated Worker thread to manage sockets proxying, main web browser thread can still not offload proxying to it, since the main thread can not synchronously pause to wait for any results. It might be possible to fake all sync call result values to succeed, and restrict to non-blocking sockets only, and perhaps extend the sockets API with custom async variants of functions if needed - but that may defeat the appeal of proxying sockets in the first place, since it would not be running unmodified code.

**Comment by juj** [comment]
This PR should be good to land now, the failure on Firefox browser tests looks intermittent.

File: src/library_websocket.js [hunk 11]
```diff
@@ -109,6 +109,7 @@ var LibraryWebSocket = {
 #endif
       return {{{ cDefine('EMSCRIPTEN_RESULT_INVALID_TARGET') }}};
     }
+    HEAPU32[urlLength>>2] = lengthBytesUTF8(socket.url);
```
Comment (kripken) [inline src/library_websocket.js]: is this change related to this PR? It didn't seem like the PR was focused on fixing library_websocket?

File: src/settings.js [hunk 12]
```diff
@@ -326,6 +326,9 @@ var SOCKET_WEBRTC = 0;
 // where addr and port are derived from the socket connect/bind/accept calls.
 var WEBSOCKET_URL = 'ws://';
 
+// If 1, the POSIX sockets API uses a proxy bridge to proxy sockets calls
```
Comment (kripken) [inline src/settings.js]: please expand on this, since "proxying" can mean "proxy from a pthread to the main thread" but I think here it means something very different?

**Comment by kripken** [comment]
Oh, I forgot to say - nice work on the docs page for this!

File: src/settings.js (hunk 12 above)
Comment (jakogut) [reply src/settings.js]: I think it's unfortunate ambiguity created by using the appropriate word for similar things in different places. Maybe we could agree on different nomenclature to avoid the ambiguity, such as POSIX sockets _bridge_?

Comment (juj) [reply src/settings.js]: ok, updated

File: src/library_websocket.js (hunk 11 above)
Comment (juj) [reply src/library_websocket.js]: Hmm no that's a separate fix that I did while authoring this.

Comment (kripken) [reply src/library_websocket.js]: Please separate it into another PR then - I worry about confusion in bisection if we end up with a regression here.

File: system/include/emscripten/posix_socket.h [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+extern EMSCRIPTEN_RESULT emscripten_init_websocket_to_posix_socket_bridge(const char *bridgeUrl);
+
+#ifdef __cplusplus
+}
+#endif
```
Comment (kripken) [inline system/include/emscripten/posix_socket.h]: why did `gl.symbols` change?

File: system/lib/libc-sockets.symbols [hunk 14]
```diff
@@ -0,0 +1,23 @@
+         U __errno_location
```
Comment (kripken) [inline system/lib/libc-sockets.symbols]: if these are for a new library, perhaps we can avoid a .symbols file, that is, to opt-in to get it the `-s` option would link it in?

File: system/lib/libc-wasm.symbols [hunk 15]
```diff
@@ -1,12 +1,31 @@
 # Auto-generated by tools/update_symbols.py.  DO NOT EDIT.
+         U __addtf3
```
Comment (kripken) [inline system/lib/libc-wasm.symbols]: were the changes here made by that automatic tool? odd they are different then, as I'm not sure why this PR would change them.

cc @sbc100 who wrote that tool.

File: tools/system_libs.py [hunk 16]
```diff
@@ -618,6 +625,28 @@ def get_depends(self):
     return depends
 
 
+class libsockets(MuslInternalLibrary, MTLibrary):
```
Comment (quantum5) [inline tools/system_libs.py]: If you are always going to link `libsockets` or `libsockets_proxy`, have you considered making them different versions of the same library?

File: tools/system_libs.py [hunk 17]
```diff
@@ -1171,6 +1200,11 @@ def add_library(lib):
     force_include.add('liblsan_rt_wasm')
     add_library(system_libs_map['liblsan_rt_wasm'])
 
+  if shared.Settings.PROXY_POSIX_SOCKETS:
+    add_library(system_libs_map['libsockets_proxy'])
+  else:
+    add_library(system_libs_map['libsockets'])
```
Comment (quantum5) [inline tools/system_libs.py]: Is it a good idea to always link sockets? This will include the libraries unconditionally. If you want the library to be used when needed, use a symbol file.

File: system/include/emscripten/posix_socket.h (hunk 13 above)
Comment (juj) [reply system/include/emscripten/posix_socket.h]: I ran through the symbol update tool, which did changes to this file. Looks like old changes have not been committed. I'll drop the changes from this PR.

File: system/lib/libc-sockets.symbols (hunk 14 above)
Comment (juj) [reply system/lib/libc-sockets.symbols]: Dropped symbols file.

File: system/lib/libc-wasm.symbols (hunk 15 above)
Comment (juj) [reply system/lib/libc-wasm.symbols]: Removed the changes made by the tool.

File: tools/system_libs.py (hunk 16 above)
Comment (juj) [reply tools/system_libs.py]: I looked at that machinery, but it was not obvious to me how that works.

File: tools/system_libs.py (hunk 17 above)
Comment (juj) [reply tools/system_libs.py]: libc with sockets has been linked in unconditionally from a long time ago (musl has sockets code built-in), this PR does not seek to change that model. Including the libraries unconditionally does not add to code size if the target project does not use any of the sockets symbols.

(I do not oppose moving to requiring e.g. `-lsocket` to get sockets code, but don't want to work on changing that model here)

File: tools/system_libs.py (hunk 16 above)
Comment (kripken) [reply tools/system_libs.py]: @quantum5 maybe some docs would help explain how to use features like different library versions?

Comment (quantum5) [reply tools/system_libs.py]: Added some explanation in #9215

**Comment by hortonelectric** [comment]
@juj is this resolved yet? @quantum5

**Comment by juj** [comment]
Brought this PR up to latest now, looks like all is passing nicely. This should be ready to land now.

> @juj There appears to be a problem with a program I'm attempting to compile that uses`getpeername`. Upon being called, the proxy server reports: `Received too small sockets call message! size: 0 bytes, expected at least 8 bytes`, and upon investigating, the size passed to `emscripten_websocket_send_binary` is negative. I'm looking into it.

Thanks for taking a look at this. I had a detour for a while working on other things, and now coming back to this one. I think if there's a bug in this one, let's look at that as a followup, preferably with a test case? (if you're still interested/active on this?)

File: src/library_websocket.js (hunk 11 above)
Comment (juj) [reply src/library_websocket.js]: Landed that as https://github.com/emscripten-core/emscripten/pull/9896

**Comment by juj** [comment]
> I think we must improve our testing before landing this - we currently don't run any sockets code at all on CI, and this adds tests only to there. How about adding sockets to CI (`.circleci/config.yml`) to run at least this test and hopefully some others that are robust enough?

Had a look at the test suite, but I am not certain what to add where. If you want to locally run these tests, you need to use python 2, as the existing websockify code in the sockets suite is too old to work in python 3, and you also need to do `npm install ws` once in Emscripten root directory to install node.js web server.

**Comment by juj** [comment]
> @juj A library I'm trying to use appears to `poll` a socket before reading from it, and enters an infinite loop waiting for data. If I understand correctly, a `poll` would not currently be proxied, would it?

Yeah, `poll` and `select` and others are currently not proxied.

**Comment by kripken** [comment]
For the tests, I think another job like this one, and right after it: https://github.com/emscripten-core/emscripten/blob/7e93e0398b0c761d5f16ee507ad13016a9503443/.circleci/config.yml#L226 "run socket tests", as the name, maybe.

**Comment by juj** [comment]
> For the tests, I think another job like this one, and right after it:

Thanks, good pointer, followed that pattern and added a sockets tests suite, which seems to pass nicely.

**Comment by jakogut** [comment]
> Brought this PR up to latest now, looks like all is passing nicely. This should be ready to land now.
> 
> > @juj There appears to be a problem with a program I'm attempting to compile that uses`getpeername`. Upon being called, the proxy server reports: `Received too small sockets call message! size: 0 bytes, expected at least 8 bytes`, and upon investigating, the size passed to `emscripten_websocket_send_binary` is negative. I'm looking into it.
> 
> Thanks for taking a look at this. I had a detour for a while working on other things, and now coming back to this one. I think if there's a bug in this one, let's look at that as a followup, preferably with a test case? (if you're still interested/active on this?)

Yeah, I think I did end up fixing this one, and I can follow up with a PR and test case after this lands.

**Comment by jakogut** [comment]
> I'm unsure what the code under `tools/websocket_*` is. How does that relate to the other code (in system/)?

This is the server application that proxies socket calls made by clients over websockets to the native system. I've successfully used this, for example, to play a video stream over HTTP using FFplay compiled to WebAssembly.

**Comment by juj** [comment]
Ping @kripken for approval?

File: .circleci/config.yml [hunk 18]
```diff
@@ -237,6 +237,39 @@ commands:
...
+            EMTEST_DETECT_TEMPFILE_LEAKS: "0"
+            # --no-sandbox becasue we are running as root and chrome requires
+            # this flag for now: https://crbug.com/638180
+            CHROME_FLAGS_BASE: "--no-first-run -start-maximized --no-sandbox --use-gl=swiftshader --user-data-dir=/tmp/chrome-emscripten-profile"
+            CHROME_FLAGS_HEADLESS: "--headless --remote-debugging-port=1234"
+            CHROME_FLAGS_WASM: "--enable-features=WebAssembly --enable-features=SharedArrayBuffer --disable-features=WebAssemblyTrapHandler --js-flags=\"--experimental-wasm-threads --harmony-sharedarraybuffer --no-wasm-disable-structured-cloning\""
+            CHROME_FLAGS_NOCACHE: "--disk-cache-dir=/dev/null --disk-cache-size=1 --media-cache-size=1 --disable-application-cache --incognito"
+          command: |
+            export EMTEST_BROWSER="/usr/bin/google-chrome $CHROME_FLAGS_BASE $CHROME_FLAGS_HEADLESS $CHROME_FLAGS_WASM $CHROME_FLAGS_NOCACHE"
+            python tests/runner.py sockets
```
Comment (kripken) [inline .circleci/config.yml]: I wish there was a way to share this code instead of copying it, but I'm not sure if there is in circleci...

File: .circleci/config.yml [hunk 19]
```diff
@@ -308,6 +341,10 @@ jobs:
     executor: bionic
     steps:
       - test-chrome
+  test-sockets-chrome:
```
Comment (kripken) [inline .circleci/config.yml]: This may as well use upstream (build-upstream instead of build, etc.), so that it doesn't get removed when we remove fastcomp. Also, upstream tests run faster.

**Comment by juj** [comment]
Test failures are unrelated, looks like LLVM has added indentation checking to freetype code.

**Comment by hortonelectric** [comment]
Hey you guys, I'm working on something that uses POSIX, and I am still getting  `missing function: sigaltstack` . I am doing `-s PROXY_POSIX_SOCKETS=1`

I'm not sure if I can actually get around this or not.

**Comment by hortonelectric** [comment]
`error: undefined symbol: emscripten_websocket_send_binary ` ...hmmm..?
